src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs

```diff
@@ -125,5 +147,110 @@ public static partial class JsonSerializer
 
             // We don't add the value to the dictionary here because we need to support the read-ahead functionality for Streams.
         }
+
+        private static MetadataPropertyName GetMetadataProperty(ReadOnlySpan<byte> propertyName, ref ReadStack state, ref Utf8JsonReader reader)
+        {
+            if (state.Current.ShouldHandleReference)
+            {
+                ThrowHelper.ThrowJsonException_MetadataReferenceObjectCannotContainOtherProperties();
+            }
+
+            MetadataPropertyName metadata = GetMetadataPropertyName(propertyName, ref state, ref reader);
+            state.Current.MetadataProperty = metadata;
+            return metadata;
+        }
+
+        private static void ResolveMetadataOnDictionary(MetadataPropertyName metadata, ref ReadStack state)
+        {
+            bool isPreserved = state.Current.IsProcessingProperty(ClassType.Dictionary) ?
+                state.Current.DictionaryPropertyIsPreserved : state.Current.IsPreserved;
+
+            if (metadata == MetadataPropertyName.Id)
+            {
+                // Check we are not parsing into immutable.
+                if (state.Current.TempDictionaryValues != null)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataCannotParsePreservedObjectIntoImmutable(state.Current.JsonPropertyInfo.DeclaredPropertyType);
+                }
+
+                if (state.Current.KeyName != null || isPreserved)
```

Comment:
It might be a bad idea to rely on KeyName being null or not, maybe we can instead use a flag in the state that tells us if the dictionary has seen any key before.